File: src/core/lexer.rs
use crate::string_to_static;
use std::fmt::{Display, Formatter, Error};

pub struct Lexer {
  code: &'static str
}

enum Tokens {
  Node,
  //Keyword,
  String,
  //Number,
}

pub struct Token {
  token: Tokens,
  value: &'static str
}

impl Lexer {
  pub fn new(code: &'static str) -> Lexer {
    Lexer { code, }
  } 
  pub fn lexer(&self) -> Vec<Token> {
    let mut state: &str = "";
    let mut tokens: Vec<Token> = vec![];
    let mut tmp: Vec<&str> = vec![];
    let split: Vec<&str> = self.code.split("").collect();
    for char in split {
      if ["{", "}", "(", ")"].contains(&char) {
        tokens.push(Token { token: Tokens::Node, value: char });
      } else if char == "\"" {
        tmp.push(char);
        if state == "STRING" {
          state = "";
          tokens.push( Token { token: Tokens::String, value: string_to_static(tmp.join("")) });
          tmp.clear();
        } else {
          state = "STRING";
        }
      } else if char == " " && tmp.len() > 0 {
        if state == "STRING" {
          tmp.push(char);
        } else {
          state = "";
          tokens.push( Token { token: Tokens::String, value: string_to_static(tmp.join("")) });
          tmp.clear();
        }
      } else {
        tmp.push(char);
      }
    }
    tokens
  }
}

impl Display for Token {
  fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), Error> {
    write!(f, "{{ token: ").unwrap();
    match self.token {
      Tokens::Node => write!(f, "Node").unwrap(),
      Tokens::String => write!(f, "String").unwrap()
    };
    write!(f, ", value: \"{}\" }}", self.value).unwrap();
    Ok(())
  }
}

File: src/core/lexer.ts
// Tokens list
export enum Tokens {
  Node = 'Node',
  String = 'String',
  Word = 'Word',
}

// Token Node type
export type Node = '(' | ')' | '{' | '}';

// Token interface
export interface Token {
  token: Tokens,
  value: Node | string,
}

export class Lexer {
  private readonly code: string;
  constructor(content: string) {
    // Formatting content
    this.code = content
      .split(/\r?\n/g)
      .map((line: string) => line.trim())
      .join('');
  }
  public lexer(): Token[] {
    let state: string = '';
    // Container variable contains processed tokens
    const container: Token[] = [];
    // Tmp variable contains temporary code chars that has been collected by tokenizer and which will be pushed to container
    const tmp: string[] = [];

    for (const char of this.code) {
      if (['(', ')', '{', '}'].includes(char)) {
        // Rechecking if tmp variable isn't empty before processing Node char
        if (tmp.length > 0) {
          state = '';
          container.push({ token: Tokens.Word, value: tmp.join('').trim() });
          tmp.splice(0, tmp.length);
        }
        container.push({ token: Tokens.Node, value: char as Node, });
      } else if (char === '"') {
        tmp.push(char);
        if (state === Tokens.String) {
          state = '';
          container.push({ token: Tokens.String, value: tmp.join('').trim() });
          tmp.splice(0, tmp.length);
        } else {
          state = Tokens.String;
        }
      } else if (char === ' ' && tmp.length > 0) {
        if (state === Tokens.String) {
          tmp.push(char);
        } else {
          state = '';
          container.push({ token: Tokens.Word, value: tmp.join('').trim() });
          tmp.splice(0, tmp.length);
        }
      } else {
        tmp.push(char);
      }
    }
    // Removing empty tokens from container
    return container.filter((token: Token) => token.value.length > 0);
  }
}